pkg/export/export_test.go

```diff
@@ -332,64 +330,179 @@ type testMetricService struct {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+
+	exporterOpts := ExporterOpts{DisableAuth: true}
+	exporterOpts.Default()
+	e, err := New(ctx, log.NewJSONLogger(log.NewSyncWriter(os.Stderr)), nil, exporterOpts, NopLease())
+	if err != nil {
+		t.Fatalf("Create exporter: %s", err)
 	}
+	e.SetLabelsByIDFunc(func(storage.SeriesRef) labels.Labels {
+		return labels.FromStrings("location", "us-central1-c")
+	})
+
+	metricServer := testMetricService{}
+	e.newMetricClient = func(_ context.Context, _ ExporterOpts) (metricServiceClient, error) {
+		return &metricServer, nil
+	}
+	// Sends a sample with no labels. The project label is automatically added by the
+	// exporter.
+	sendSample := func() {
+		e.Export(nil, []record.RefSample{{Ref: 1, T: int64(0), V: float64(0)}}, nil)
+	}
+	// Tests all samples have the correct project ID label value.
+	testSamples := func(expectedProjectID string, expectedSampleCount int) {
+		var err error
+		pollErr := wait.PollUntilContextCancel(ctx, batchDelayMax, false, func(_ context.Context) (bool, error) {
+			switch len(metricServer.samples) {
+			case 0:
+				err = errors.New("no samples sent")
+				return false, nil
+			case expectedSampleCount:
+				// Good.
+			default:
+				// Sometimes there's a small delay from the thread that sends the new
+				// samples, so let's wait.
+				err = fmt.Errorf("expected %d samples but got %d", expectedSampleCount, len(metricServer.samples))
+				return false, nil
+			}
+
+			for _, sample := range metricServer.samples {
+				projectID := sample.Resource.Labels[KeyProjectID]
+				if projectID != expectedProjectID {
+					err = fmt.Errorf("expected project ID %q but got %q", expectedProjectID, projectID)
+					return false, nil
+				}
+			}
+
+			return true, nil
+		})
+		if pollErr != nil {
+			if wait.Interrupted(pollErr) && err != nil {
+				pollErr = err
+			}
+			t.Fatalf("did not get samples: %s", pollErr)
+		}
+	}
+	sendAndTestSamples := func(expectedProjectID string) {
+		// Send two samples to ensure both have correct labels.
+		sendSample()
+		sendSample()
+		sendSample()
+		testSamples(expectedProjectID, 3)
+		metricServer.clear()
+	}
+
+	// In our Prometheus fork, GCM is executed before the reloader in the run group.
+	go func() {
+		if err := e.Run(); err != nil {
+			t.Errorf("Run exporter: %s", err)
+		}
+	}()
+
+	if err := e.ApplyConfig(&config.Config{}, &ExporterOpts{ProjectID: "project-test"}); err != nil {
+		t.Fatalf("Initial apply: %s", err)
+	}
+	sendAndTestSamples("project-test")
+
+	if err := e.ApplyConfig(&config.Config{}, &ExporterOpts{ProjectID: "project-abc"}); err != nil {
+		t.Fatalf("Initial apply: %s", err)
+	}
+	sendAndTestSamples("project-abc")
+
+	if err := e.ApplyConfig(&config.Config{}, &ExporterOpts{ProjectID: "project-xyz"}); err != nil {
+		t.Fatalf("Initial apply: %s", err)
+	}
+	sendAndTestSamples("project-xyz")
+}
+
+func TestDisabledExporter(t *testing.T) {
```

Comment:
question: What's being tested in this function?

Reply:
Added more to the test to make it clearer.